sway/scratchpad.c

```diff
@@ -0,0 +1,173 @@
+#define _XOPEN_SOURCE 700
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include "sway/scratchpad.h"
+#include "sway/input/seat.h"
+#include "sway/tree/arrange.h"
+#include "sway/tree/container.h"
+#include "sway/tree/view.h"
+#include "sway/tree/workspace.h"
+#include "list.h"
+#include "log.h"
+
+void scratchpad_add_container(struct sway_container *con) {
+	if (!sway_assert(!con->scratchpad, "Container is already in scratchpad")) {
+		return;
+	}
+	con->scratchpad = true;
+	list_add(server.scratchpad, con);
+
+	struct sway_container *parent = con->parent;
+	container_set_floating(con, true);
+	container_remove_child(con);
+	arrange_windows(parent);
+
+	struct sway_seat *seat = input_manager_current_seat(input_manager);
+	seat_set_focus(seat, seat_get_focus_inactive(seat, parent));
+}
+
+void scratchpad_remove_container(struct sway_container *con) {
+	if (!sway_assert(con->scratchpad, "Container is not in scratchpad")) {
+		return;
+	}
+	con->scratchpad = false;
+	for (int i = 0; i < server.scratchpad->length; ++i) {
+		if (server.scratchpad->items[i] == con) {
+			list_del(server.scratchpad, i);
+			break;
+		}
+	}
+}
+
+/**
+ * Show a single scratchpad container.
+ * The container might be visible on another workspace already.
+ */
+static void scratchpad_show(struct sway_container *con) {
+	struct sway_seat *seat = input_manager_current_seat(input_manager);
+	struct sway_container *ws = seat_get_focus(seat);
+	if (ws->type != C_WORKSPACE) {
+		ws = container_parent(ws, C_WORKSPACE);
+	}
+
+    // If the current con or any of its parents are in fullscreen mode, we
+    // first need to disable it before showing the scratchpad con.
+	if (ws->sway_workspace->fullscreen) {
+		view_set_fullscreen(ws->sway_workspace->fullscreen, false);
+	}
+
+	// Show the container
+	if (con->parent) {
+		container_remove_child(con);
+	}
+	container_add_child(ws->sway_workspace->floating, con);
+
+	// Make sure the container's center point overlaps this workspace
+	double center_lx = con->x + con->width / 2;
+	double center_ly = con->y + con->height / 2;
+
+	struct wlr_box workspace_box;
+	container_get_box(ws, &workspace_box);
+	if (!wlr_box_contains_point(&workspace_box, center_lx, center_ly)) {
+		// Maybe resize it
+		if (con->width > ws->width || con->height > ws->height) {
+			// TODO: Do this properly once we can float C_CONTAINERs
+			if (con->type == C_VIEW) {
+				view_init_floating(con->sway_view);
+				arrange_windows(con);
+			}
+		}
+
+		// Center it
```

Comment:
This does not match i3's behavior. For views that were tiling that are added to the scratchpad, this does match. For views that were floating (or moved while in the scratchpad), the location is retained relative to a pseudo output.

https://github.com/i3/i3/blob/next/src/scratchpad.c#L248
https://github.com/i3/i3/blob/next/src/floating.c#L926

(I'm fine with this being changed in a follow-up PR)

Reply:
> For views that were floating (or moved while in the scratchpad), the location is retained relative to a pseudo output.

Just confirming, the behaviour you're describing is what the PR currently does, but it should center the container every time it's displayed?

Reply:
So currently this PR centers the view if not in bounds of the output.

In i3, if you have two outputs of the same resolution, the view will always be at the same (x, y) relative to the top-left of the output.

For outputs of different resolutions, the views may extend over the bounds of the output. I'm not exactly sure how the views are being translated on the outputs. I think the second link contains the  function doing the translation.